packages/swapper/src/swappers/RelaySwapper/utils/getTrade.ts

```diff
@@ -0,0 +1,365 @@
+      }),
+    )
+  }
+
+  const protocolAssetId = relayTokenToAssetId(relayToken)
+
+  const protocolAsset = relayTokenToAsset(relayToken, deps.assetsById)
+
+  const steps = swapSteps.map((quoteStep): TradeQuoteStep => {
+    const isCrossChain = sellAsset.chainId !== buyAsset.chainId
+
+    const appFeesBaseUnit = (() => {
+      // Handle app fees based on cross-chain status
+      if (
+        quote.fees?.app?.amount &&
+        quote.fees?.app?.currency &&
+        isValidRelayToken(quote.fees?.app?.currency)
+      ) {
+        const appFeesAsset = quote.fees?.app?.currency
+          ? relayTokenToAsset(quote.fees.app.currency, deps.assetsById)
+          : undefined
+
+        // @TODO: we might need to change this logic when solana and BTC are supported
+        const isNativeCurrencyInput =
+          isNativeEvmAsset(sellAsset.assetId) && sellAsset.chainId === appFeesAsset?.chainId
+
+        // For cross-chain: always add back app fees
+        // For same-chain: only add back if input is native currency
+        if (isCrossChain || isNativeCurrencyInput) {
+          return quote.fees.app.amount
+        }
+      }
+
+      // cross-chain or same-chain with native currency as input are not applicable
+      return '0'
+    })()
+
+    const relayerFeesBuyAssetBaseUnit = (() => {
+      const relayerFeeRelayToken = quote.fees?.relayer?.currency
+      if (relayerFeeRelayToken && isValidRelayToken(relayerFeeRelayToken)) {
+        const relayerFeesAsset = relayTokenToAsset(relayerFeeRelayToken, deps.assetsById)
+        const relayerFeeAmount = quote.fees?.relayer?.amount ?? '0'
+
+        // If fee is already in buy asset, return as is
+        if (relayerFeesAsset?.assetId === buyAsset.assetId) {
+          return relayerFeeAmount
+        }
+
+        // if fee is in sell asset, convert to buy asset
+        if (relayerFeesAsset?.assetId === sellAsset.assetId) {
+          return convertPrecision({
+            value: relayerFeeAmount,
+            inputExponent: relayerFeesAsset.precision,
+            outputExponent: buyAsset.precision,
+          })
+            .times(rate)
+            .toFixed(0)
+        }
+
+        // If fee is in a different asset, convert to buy asset
+        const feeAmountUsd = quote.fees?.relayer?.amountUsd ?? '0'
+        const buyAssetUsd = quote.details?.currencyOut?.amountUsd ?? '0'
+        const buyAssetAmountBaseUnit = quote.details?.currencyOut?.amount ?? '1'
+
+        if (feeAmountUsd && buyAssetUsd && buyAssetAmountBaseUnit) {
+          // Calculate the rate: (buyAssetAmount / buyAssetUsd) gives us "buy asset per USD"
+          // Then multiply by feeAmountUsd to get the equivalent buy asset amount
+          const buyAssetCryptoBaseUnitPerUsd = bnOrZero(buyAssetAmountBaseUnit).div(buyAssetUsd)
+          const buyAssetFeesCryptoBaseUnit = bnOrZero(feeAmountUsd).times(
+            buyAssetCryptoBaseUnitPerUsd,
+          )
```

Comment:
That's stupid and we loose precision but I couldn't find another way to calculate that

Fee asset can be something else than buyAsset and sellAsset so we don't have the exact price of this asset, I get the rate by using the USD value of the current fees relay returns to us, then convert it to the sellAsset using it's own rate

Reply:
Keeping this open to double-check this at runtime but doesn't look shocking to me as the sole purpose of this is to be added back and get the amount before fees

Reply:
I guess nothing shocking then! closing this one